cilium/CFP-30235-xds-adapter.md

```diff
@@ -0,0 +1,213 @@
+2. Even if there were a feedback loop, spillover to other zones benefits from
+   some form of centralized orchestration to avoid a thundering herd problem.
+3. All changes to endpoint routing currently have to go through Services or
+   EndpointSlices, and by extension, the Kubernetes API Server. It would be very
+   expensive to implement incremental weighting/spillover adjustments with that
+   overhead. For example, adjusting the weight of any individual endpoint would
+   require a write to the EndpointSlice API which would then need to be
+   distributed to all consumers of that API. That can be problematic from a
+   scalability perspective when there are frequent updates paired with a large
+   number of endpoints and/or nodes.
+
+I believe that xDS is uniquely positioned to address these limitations.
+[LRS](https://www.envoyproxy.io/docs/envoy/latest/api-v3/service/load_stats/v3/lrs.proto)
+(Load Reporting Service) provides a straightforward way to provide load reports
+to a centralized control plane. Similarly, [Delta
+xDS](https://www.envoyproxy.io/docs/envoy/latest/configuration/overview/xds_api#delta-endpoints)
+enables xDS to distribute incremental updates that can be smaller in scope than
+comparable Kubernetes APIs. All of this could be combined to avoid adding
+additional load on the Kubernetes API Server.
+
+#### Endpoints From Other Sources
+
+In some cases, it’s useful to be able to route to endpoints from other sources,
+such as other Kubernetes clusters or outside of Kubernetes altogether. This can
+be particularly relevant when implementing Multi-Cluster Service routing.
+Implementations today often rely on one of the following approaches:
+
+1. Mirror endpoints from other clusters into Kubernetes by creating custom
+   Services and EndpointSlices.
+2. Introduce Gateways on the edges of clusters that can be used to route
+   requests from other clusters.
+
+With xDS support, it could be possible to have a centralized xDS control plane
+that was familiar with endpoints in multiple sources, providing a potentially
+more efficient alternative to either of the approaches commonly used today.
+
+## Proposal
+
+This work will be divided into several phases. Splitting this work into phases
+will enable us to focus on smaller deliverables that each independently provide
+value to the project.
+
+### 1. Common Interface for Data Sources
+Before we get started on an xDS adapter for Cilium, we'll need to build a common
+interface for interacting with data sources (Kubernetes, KVStore, and xDS). This
+interface will enable the majority of the codebase to be entirely unaware of
+which data sources are in use.
+
+A key part of defining this interface will involve determining a precedence
+order when more than one data source is connected. When any conflicts occur,
+for example if Services from different sources have the same IP, the following
+order will be used for precedence:
+
+1. Kubernetes
```

Comment:
`0. local API` - It's possible to add endpoints and services using the local API.

Reply:
yes, I think that the first part of this effort needs to be to understand the existing internal model for endpoints and endpoint groups and how well it aligns with the other models (including xDS).

Reply:
+1, phase 1 in this proposal will be to define a common interface that can be used for all Service and Endpoint data sources. I think that _mostly_ exists, but it would be good to ensure that _everything_ is going through the same shared path here with an interface we all agree on before going any further. I also added the local API to the list as suggested by @aanm above.